util/fpga/bitstream_bisect.py

```diff
@@ -0,0 +1,406 @@
+            rich.print(
+                rich.panel.Panel(rich.markup.escape(proc.stdout),
+                                 title="git stdout"))
+            rich.print(
+                rich.panel.Panel(rich.markup.escape(proc.stderr),
+                                 title="git stderr"))
+        return (proc.returncode, proc.stdout, proc.stderr)
+
+    def _git_no_capture_output(self, *args: str) -> int:
+        command = [self._git_binary] + list(args)
+        command_str = " ".join(command)
+        command_str = rich.markup.escape(command_str)
+        self._console.rule(
+            "Streaming output from [bold]{}".format(command_str))
+        proc = subprocess.run(["git"] + list(args), env=self._env)
+        self._console.rule("Git exited with status {}".format(proc.returncode))
+        self._console.print()
+        return proc.returncode
+
+
+class BisectSession:
+
+    def __init__(self,
+                 git: GitControllerABC,
+                 cache_keys: Set[CommitHash],
+                 visualize: bool = False):
+        self._git = git
+        self._cache_keys: Set[CommitHash] = cache_keys
+        self._console = rich.console.Console()
+        self._visualize = visualize
+
+    def run(self,
+            good_commit: CommitHash,
+            bad_commit: CommitHash,
+            fast_command: List[str],
+            slow_command: Optional[List[str]] = None) -> BisectLog:
+        """Run an unattended bisect session in one or two phases.
+
+        This method is conceptually equivalent to `git bisect run`. Unlike plain
+        `git bisect`, it runs in two phases. The first phase only operates on
+        cached bitstreams. This will only get us so far, so it optionally can
+        fall back to a slower command, e.g. one that builds the bitstreams it
+        tests.
+
+        Args:
+          good_commit:
+            Commit where `fast_command` and `slow_command` are known to succeed.
+          bad_commit:
+            Commit where `fast_command` and `slow_command` are known to fail.
+          fast_command:
+            A list of strings that represent a git-bisect script and args. These
+            are passed to `git-bisect run` in the first phase.
+          slow_command:
+            Just like `fast_command`, but for the second phase.  If this arg is
+            None, we skip the second phase.
+
+        Returns:
+          Parsed output from `git bisect log`, run after the final bisect phase.
+        """
+
+        self._git.bisect_start()
+        self._git.bisect_good(good_commit)
+        self._git.bisect_bad(bad_commit)
+
+        # Skip commits that aren't in the bitstream cache.
+        for commit, _ in self._git.bisect_view():
+            if commit not in self._cache_keys:
+                self._git.bisect_skip(commit)
+
+        self._maybe_visualize("Only testing commits with cached bitstreams")
+        bisect_log = self._git.bisect_run(*fast_command)
```

Comment:
Maybe: Run the fast command with the BITSTREAM=$commit environment variable

Reply:
I like this from a belt-and-suspenders angle. It's not trivial to implement because `git bisect run` is responsible for invoking the fast command, so we'd have to wrap the command in some runtime-evaluated logic.

I'll punt on this as well because the PR is already huge, but I'll file an issue because I think it's worth doing.